// prettier-plugin-json-formats/create-plugin/parser/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportTrailingCommas = exports.supportSingleQuotes = exports.supportNumberConstants = exports.supportMultilineStrings = exports.supportLaxNumbers = exports.supportIdentifiers = exports.supportHexadecimalNumbers = exports.supportComments = void 0;
const flags_1 = require("../flags");
function createFlagSwitch(flag) {
    return context => (context.flags & flag) != 0;
}
exports.supportComments = createFlagSwitch(flags_1.JsonFlags.CommentsAllowed);
exports.supportHexadecimalNumbers = createFlagSwitch(flags_1.JsonFlags.HexadecimalNumberAllowed);
exports.supportIdentifiers = createFlagSwitch(flags_1.JsonFlags.IdentifierKeyNamesAllowed);
exports.supportLaxNumbers = createFlagSwitch(flags_1.JsonFlags.LaxNumberParsingAllowed);
exports.supportMultilineStrings = createFlagSwitch(flags_1.JsonFlags.MultiLineStringAllowed);
exports.supportNumberConstants = createFlagSwitch(flags_1.JsonFlags.NumberConstantsAllowed);
exports.supportSingleQuotes = createFlagSwitch(flags_1.JsonFlags.SingleQuotesAllowed);
exports.supportTrailingCommas = createFlagSwitch(flags_1.JsonFlags.TrailingCommasAllowed);
