// prettier-plugin-json-formats/angular-cli.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.angularCliPlugin = void 0;
const create_plugin_1 = require("./create-plugin");
const plugin = (0, create_plugin_1.createJsonPlugin)({
    language: { name: 'angular-cli' },
    printer: 'json',
    flags: create_plugin_1.JsonFlags.CommentsAllowed,
    parserFlags: create_plugin_1.JsonFlags.CommentsAllowed,
    modifier: (0, create_plugin_1.combine)((0, create_plugin_1.sortObjectProperties)([
        '$schema',
        'version',
        'cli',
        'newProjectRoot',
        'projects',
        'schematics',
    ]), (0, create_plugin_1.replacePropertyValue)('projects', (0, create_plugin_1.combine)((0, create_plugin_1.sortObjectProperties)((projects, options) => {
        const keys = (0, create_plugin_1.getPropertyKeys)(projects).sort();
        const sortAtTheTop = new Set(options.angularCliTopProjects);
        const sortAtTheBottom = new Set(options.angularCliBottomProjects);
        return [
            ...keys.filter(key => sortAtTheTop.has(key)),
            ...keys.filter(key => !sortAtTheTop.has(key) && !sortAtTheBottom.has(key)),
            ...keys.filter(key => sortAtTheBottom.has(key)),
        ];
    }), (0, create_plugin_1.replacePropertyValues)((0, create_plugin_1.combine)((0, create_plugin_1.renameProperty)('targets', 'architect'), (0, create_plugin_1.sortObjectProperties)([
        'projectType',
        'root',
        'sourceRoot',
        'architect',
        'i18n',
        'schematics',
    ]), (0, create_plugin_1.replacePropertyValue)('architect', (0, create_plugin_1.combine)((0, create_plugin_1.sortObjectProperties)(), (0, create_plugin_1.replacePropertyValues)((0, create_plugin_1.combine)((0, create_plugin_1.sortObjectProperties)([
        'builder',
        'options',
        'configurations',
        'schematics',
    ]), (0, create_plugin_1.replacePropertyValues)((0, create_plugin_1.deepSortObjectProperties)()))))), (0, create_plugin_1.replacePropertyValue)('i18n', (0, create_plugin_1.deepSortObjectProperties)()), (0, create_plugin_1.replacePropertyValue)('schematics', (0, create_plugin_1.deepSortObjectProperties)()))))), (0, create_plugin_1.replacePropertyValue)('schematics', (0, create_plugin_1.deepSortObjectProperties)())),
});
exports.angularCliPlugin = {
    ...plugin,
    options: {
        angularCliTopProjects: {
            category: 'Angular CLI',
            default: [{ value: [] }],
            type: 'string',
            array: true,
            description: 'Keys of projects to sort at the top',
        },
        angularCliBottomProjects: {
            category: 'Angular CLI',
            default: [{ value: [] }],
            type: 'string',
            array: true,
            description: 'Keys of projects to sort at the bottom',
        },
    },
};
