// prettier-plugin-json-formats/create-plugin/flags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonFlags = void 0;
/**
 * Mode used for parsing and printing JSON
 */
var JsonFlags;
(function (JsonFlags) {
    JsonFlags[JsonFlags["Strict"] = 0] = "Strict";
    JsonFlags[JsonFlags["CommentsAllowed"] = 1] = "CommentsAllowed";
    JsonFlags[JsonFlags["SingleQuotesAllowed"] = 2] = "SingleQuotesAllowed";
    JsonFlags[JsonFlags["IdentifierKeyNamesAllowed"] = 4] = "IdentifierKeyNamesAllowed";
    JsonFlags[JsonFlags["TrailingCommasAllowed"] = 8] = "TrailingCommasAllowed";
    JsonFlags[JsonFlags["HexadecimalNumberAllowed"] = 16] = "HexadecimalNumberAllowed";
    JsonFlags[JsonFlags["MultiLineStringAllowed"] = 32] = "MultiLineStringAllowed";
    JsonFlags[JsonFlags["LaxNumberParsingAllowed"] = 64] = "LaxNumberParsingAllowed";
    JsonFlags[JsonFlags["NumberConstantsAllowed"] = 128] = "NumberConstantsAllowed";
    JsonFlags[JsonFlags["Default"] = 0] = "Default";
    JsonFlags[JsonFlags["Loose"] = 255] = "Loose";
    JsonFlags[JsonFlags["Json"] = 0] = "Json";
    JsonFlags[JsonFlags["Json5"] = 255] = "Json5";
    JsonFlags[JsonFlags["JsonWithComments"] = 1] = "JsonWithComments";
})(JsonFlags || (exports.JsonFlags = JsonFlags = {}));
